### 📌 정상 통화 중 '은행', '대출' 포함된 샘플을 충분히 모델에 학습
- Label == 0 (정상 통화)인데
- '은행','대출','신용' 등의 금융 키워드가 포함된 샘플 찾기

✅ 1단계: 병합된 CSV 불러오기

In [4]:
import pandas as pd

# 병합된 파일 경로
merged_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_merged.csv'

# CSV 로드
df = pd.read_csv(merged_path)

✅ 2단계: '은행', '대출', '신용' 키워드 포함 + Label=0 필터링

In [7]:
# 키워드 목록 (확장 가능)
keywords = ['은행', '대출', '신용']

# 정규표현식으로 키워드 결합
keyword_pattern = '|'.join(keywords)

# 조건: Label == 0 (정상 통화) 이고, Transcript에 키워드 포함
filtered_df = df[(df['Label'] == 0) & (df['Transcript'].str.contains(keyword_pattern, regex=True))]

# 결과 개수 확인
print(f"✅ 키워드 포함된 정상 통화 샘플 수: {len(filtered_df)}개")

✅ 키워드 포함된 정상 통화 샘플 수: 70개


✅ 3단계: 샘플 미리보기

In [10]:
# 상위 5개 예시 출력
print("\n📋 예시 샘플:")
for i, row in filtered_df.head(5).iterrows():
    print(f"- {row['Transcript'][:100]}... (Filename: {row['Filename']})")


📋 예시 샘플:
- 아이 10 아니 생각 보다 너무 올라 그래서 진작 어야 요즘 엄마 한테 애기 엄마 한테 그냥 얼마 얼마나 나오 세금 세금 우리 그니까 대출 으니까 대출 면은 500 그거 충격 아니... (Filename: KorCCViD)
- 안녕하세요. 공공은행입니다. 신용대출을 받을 수 있을까요? 본인이 맞는지 먼저 확인하겠습니다. 네. 안내 멘트가 나오면 생년월일 6자리를 눌러주세요. 네 알겠습니다. 확인 되었습니... (Filename: normal_normal_607.wav)
- 신용대출은 인터넷으로 신청할 수 있어요? 아니요 방문하셔야 합니다 아 그렇군요 아무 노겹이다 가면 되나요? 네 맞습니다 서류는 뭘 가져가면 되죠? 신분증과 소득 확인 서류 가져가시... (Filename: normal_normal_543.wav)
- 보험료 출금해 주세요. 본인 확인 후 안내 드리겠습니다. 성함을 말씀해 주시겠어요? 0000입니다. 주민번호 앞자리 말씀해 주시겠어요? 000000입니다. 나비 중이신 보험료 이체... (Filename: normal_normal_370.wav)
- 전세금으로 담보대출 상품을 알아볼 수 있나요? 전세계약을 진행하셨나요? 지금 살고 있는 집은 안 되나요? 전세담보대출은 이미 살고 계신 집은 대출 진행이 불과합니다. 그럼 전세자금... (Filename: normal_normal_92.wav)


✅ 4단계 (선택): 새 CSV로 저장

In [13]:
filtered_df.to_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/normal_with_financial_keywords.csv', index=False)
print("📁 해당 샘플을 별도 CSV로 저장했습니다.")

📁 해당 샘플을 별도 CSV로 저장했습니다.


### ✅ 목표

- Label == 1 (보이스 피싱)인데

- '은행', '대출', '신용' 등의 금융 키워드가 전혀 포함되지 않은 샘플을 찾기

✅ 1단계: 키워드 미포함 보이스 피싱 샘플 필터링

In [21]:
# 조건: Label == 1 AND 키워드 미포함
phishing_without_keywords = df[
    (df['Label'] == 1) & (~df['Transcript'].str.contains(keyword_pattern, regex=True))
]

print(f"🚨 키워드 없이도 보이스 피싱인 샘플 수: {len(phishing_without_keywords)}개")

🚨 키워드 없이도 보이스 피싱인 샘플 수: 348개


✅ 2단계: 예시 출력

In [24]:
print("\n📋 키워드 없는 보이스 피싱 샘플 예시:")
for i, row in phishing_without_keywords.head(5).iterrows():
    print(f"- {row['Transcript'][:100]}... (Filename: {row['Filename']})")


📋 키워드 없는 보이스 피싱 샘플 예시:
- 사람 전라도 광주 태장 42 고요 명동 에서 10 정도 근무 했었 200 보여 그리고 인데 저녁 사람 으십니까 면은 얼마 저희 수학 에서 현장 에서 우리 하나 통장 고요 하나 만들... (Filename: KorCCViD)
- 여기 끝내 2005 김창호 도용 사건 고요 감당 서울 중앙 지검 지능 범죄 수사 외계인 밝혀 면서 입니다 명일동 통장 불법 현장 에서 나왔 때문 본인 직접 결과 다니 아니 당하 던... (Filename: KorCCViD)
- 답답 지금 선생 본인 래요 자기 성함 아니 에요 아니 통해서 통화 어요 답답 십니다 진짜 아니 결과 아실 예요 지금 어떤 사태 지금 발생 복잡 어요 지금 굉장히 복잡 경찰 에서 아... (Filename: KorCCViD)
- 서울 중앙 지검 김진호 수사관 입니다 잠시 통화 세요 언제 실까요 다름 아니 명의 도용 결제 한도 혐의 본인 관련 사건 연료 어서 확인 연락 드린 겁니다 언제 통화 실까요 서울 중... (Filename: KorCCViD)
- 확인 농협 하나 통장 확인 경기도 광명시 무슨 입니까 통장 경우 부장 확인 말씀 드렸 개인 정보 유출 계실 겁니다 휴대폰 지금 신분증 부분 대해서 분실 도난 당한... (Filename: KorCCViD)


### 🎯 분석 요약
| 항목                              | 수치       |
| ------------------------------- | -------- |
| ✅ 키워드 포함 정상 통화 (`Label == 0`)   | 70개      |
| 🚨 키워드 없는 보이스 피싱 (`Label == 1`) | **348개**
-------------------------------------------------------
📌 의미 해석
- 🔍 보이스 피싱인데도 "은행", "대출", "신용" 같은 키워드 없이 진행되는 사례가 매우 많다! |


## !!!기존에 만든 모델에 넣어 나온 탐지 결과 csv 

✅ 1. 예측 정확도 요약 (정답률, 오답률 등)

In [34]:
import pandas as pd

# CSV 로드
df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted.csv')

# ✅ 총 개수
total = len(df)

# ✅ 정확히 맞춘 샘플 수
correct = (df['Label'] == df['PredictedLabel']).sum()
incorrect = total - correct

# ✅ 클래스별 개수
from collections import Counter
true_label_dist = Counter(df['Label'])
pred_label_dist = Counter(df['PredictedLabel'])

# ✅ 정밀도 / 재현율 / F1 계산
from sklearn.metrics import classification_report

report = classification_report(df['Label'], df['PredictedLabel'], target_names=['정상 통화', '보이스 피싱'])

# 출력
print("📊 예측 통계 요약:")
print(f"- 총 샘플 수: {total}")
print(f"- 정확히 예측한 샘플: {correct}개 ({correct/total*100:.2f}%)")
print(f"- 잘못 예측한 샘플: {incorrect}개 ({incorrect/total*100:.2f}%)")

print("\n🔎 실제 라벨 분포:", dict(true_label_dist))
print("🔍 예측 라벨 분포:", dict(pred_label_dist))

print("\n📈 상세 성능 리포트 (정밀도, 재현율, F1-score):")
print(report)

📊 예측 통계 요약:
- 총 샘플 수: 1418
- 정확히 예측한 샘플: 1310개 (92.38%)
- 잘못 예측한 샘플: 108개 (7.62%)

🔎 실제 라벨 분포: {1: 709, 0: 709}
🔍 예측 라벨 분포: {1: 807, 0: 611}

📈 상세 성능 리포트 (정밀도, 재현율, F1-score):
              precision    recall  f1-score   support

       정상 통화       0.99      0.85      0.92       709
      보이스 피싱       0.87      0.99      0.93       709

    accuracy                           0.92      1418
   macro avg       0.93      0.92      0.92      1418
weighted avg       0.93      0.92      0.92      1418



✅ 결과 분석: 모델의 이상한 점

| 항목                     | 분석                                                                    |
| ---------------------- | --------------------------------------------------------------------- |
| **예측 분포**              | `Label=1`이 709개인데 `PredictedLabel=1`은 807개 → 모델이 **보이스 피싱 쪽으로 편향**    |
| **정밀도 vs 재현율 (정상 통화)** | 정밀도 **0.99**는 높지만, 재현율이 **0.85**밖에 안 됨 → **정상인데도 피싱으로 잘못 판단**하는 비율 높음 |
| **오분류 대부분이 FP**        | 즉, **False Positive = 정상 통화를 보이스 피싱으로 잘못 판단하는 사례**가 집중됨               |
| **잘못된 확신도**            | 샘플 1219, 1220 등에서 확신도 98% 이상 → 모델이 확신까지 갖고 잘못 판단함 → **잘못 학습된 패턴 존재*
----------------------------------------------
- 📌 즉, 금융 키워드 포함 문장 또는 문맥이 어색한 문장이 정상이라도
모델은 피싱으로 학습했기 때문에 고신뢰도로 오판*  |


✅ 2. 실제 잘못 예측된 샘플들 가져오기 + 문제 파악

In [12]:
import pandas as pd

df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted.csv')

# 잘못 예측된 샘플 중 → 정상 통화를 피싱으로 오판한 것만 추출
false_positives = df[(df['Label'] == 0) & (df['PredictedLabel'] == 1)]

print(f"❌ 오분류된 정상 통화 샘플 수: {len(false_positives)}개")

# 대표적인 5개 문장 출력 + 간단한 원인 분석
suspicious_keywords = ['대출', '신용', '은행', '한도', '계좌', '비대면']

for i, row in false_positives.head(5).iterrows():
    transcript = row['Transcript']
    confidence = row['Confidence']
    matched_keywords = [kw for kw in suspicious_keywords if kw in transcript]

    print(f"\n📞 오분류 샘플 {i+1}")
    print(f"▶ 텍스트: {transcript[:200]}...")
    print(f"🔍 예측 확신도: {confidence*100:.2f}%")
    if matched_keywords:
        print(f"💡 원인 추정: '{', '.join(matched_keywords)}' 키워드 때문에 피싱으로 오판했을 가능성")
    elif len(transcript) < 30:
        print("💡 원인 추정: 텍스트가 너무 짧아서 문맥 이해 어려움")
    else:
        print("💡 원인 추정: 문맥 또는 어조가 피싱으로 학습된 것과 유사했을 가능성")

❌ 오분류된 정상 통화 샘플 수: 103개

📞 오분류 샘플 235
▶ 텍스트: 사실 아까 시간 얘기 으면 계획 세우 아요 오늘 10 시부 10 시작 니까 목사 10 시작 해서 12 12 어서 끝내 라고 10 에서 인제 시작 12 어서 끝나 식사 인제 글케 생각 인제 어디 갈려고 했었 어요...
🔍 예측 확신도: 53.90%
💡 원인 추정: 문맥 또는 어조가 피싱으로 학습된 것과 유사했을 가능성

📞 오분류 샘플 598
▶ 텍스트: 중앙 포스 누나 영수증 으면서 해서 고객 상품권 다른 영수증 하나 고요 다른 하나 셔야 돼요 고객 니까 거기 고객 그래요 면서 모르 근데 거기 난동 으니까 고객 얼굴 기억 거기 사람...
🔍 예측 확신도: 98.38%
💡 원인 추정: 문맥 또는 어조가 피싱으로 학습된 것과 유사했을 가능성

📞 오분류 샘플 1096
▶ 텍스트: 저번 제일 웃겼 내일 자일리톨 많이 냐는 질문 많이 니까 롯데 자일리톨 많이 드세요 해태 자일리톨 많이 드세요 라고 진짜 미치 어떻게 모르 해서 계속 많이 더니 사람 롯데 조금 많이 으시 롯데 많이...
🔍 예측 확신도: 85.31%
💡 원인 추정: 문맥 또는 어조가 피싱으로 학습된 것과 유사했을 가능성

📞 오분류 샘플 1219
▶ 텍스트: 안녕하세요. 공공은행입니다. 신용대출을 받을 수 있을까요? 본인이 맞는지 먼저 확인하겠습니다. 네. 안내 멘트가 나오면 생년월일 6자리를 눌러주세요. 네 알겠습니다. 확인 되었습니다 고객님. 성함이나 휴대폰 번호 알려주세요. 공공공이고 공공공 공공공 공공공입니다. 대출 한도가 다 대출건과 합산이 되어 산정됩니다. 필요하신 자금은 얼마신가요? 3,000만원 ...
🔍 예측 확신도: 98.21%
💡 원인 추정: '대출, 신용, 은행, 한도, 비대면' 키워드 때문에 피싱으로 오판했을 가능성

📞 오분류 샘플 1220
▶ 텍스트: 신용대출은 인터넷으로 신청할 수 있어요? 아니요 방문하셔야 합니다 아 그렇군요 아무 노겹이다 가면 되나요? 네 맞습니다 서류는 뭘 가져가면 되죠? 신

📊 오분류 샘플 분류 요약
| 유형           | 샘플             | 원인                                  | 조치                                        |
| ------------ | -------------- | ----------------------------------- | ----------------------------------------- |
| 🟥 키워드 오판형   | 1219, 1220     | `'대출'`, `'신용'`, `'은행'` 등 금융 키워드 과반응 | 👉 **금융 키워드 포함 정상 통화 oversampling 필요**    |
| 🟨 문맥 오판형    | 235, 598, 1096 | 문장이 어색하거나 의미 모호, **유사 피싱 어조 학습**    | 👉 **자연어 보정된 정상 문장 데이터 보강 또는 paraphrase** |
| ⚠️ 낮은 확신도 오판 | 235 (53.9%)    | 애매한 문장 → 모델도 확신 없음                  | 👉 **threshold 적용**으로 “보류” 판단 처리          |


✅ 3. 오분류 샘플 저장 (CSV)

In [21]:
false_positives.to_csv(
    'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/false_positives_label0_pred1.csv',
    index=False
)
print("📁 오분류 샘플 CSV 저장 완료")

📁 오분류 샘플 CSV 저장 완료


✅ 4. 금융 키워드 포함 정상 통화 Oversampling
- Label == 0 이면서 "대출|신용|은행"이 포함된 문장들을 df에서 추출
- 동일 문장을 여러 번 학습셋에 넣어 학습 비중을 증가시켜야 함

In [24]:
keyword_normal = df[(df['Label'] == 0) & (df['Transcript'].str.contains("대출|신용|은행|비대면|한도"))]
augmented_df = pd.concat([df, keyword_normal, keyword_normal])  # 2배 복제

✅ 5. Threshold 기반 예측 함수 적용 (즉시 사용 가능)

In [27]:
def predict_phishing_with_threshold(text, model, tokenizer, threshold=0.9):
    encoded = tokenizer(text, padding='max_length', truncation=True, max_length=128, return_tensors='pt')
    input_ids = encoded['input_ids'].to(device)
    attention_mask = encoded['attention_mask'].to(device)
    token_type_ids = encoded['token_type_ids'].to(device)

    with torch.no_grad():
        output = model(input_ids, attention_mask, token_type_ids)
        phishing_score = output[0][1].item()

    pred_label = 1 if phishing_score >= threshold else 0
    return pred_label, phishing_score

### 📌 최종 정리: 내 모델이 잘못 예측한 이유
- 금융 키워드 → 너무 자주 등장한 피싱 샘플로 인해 편향된 학습

- 문장이 어색하거나 짧은 경우 → 피싱과 유사하다고 판단

- 결과적으로 False Positive가 많음 → 실제 사용자 혼란 초래 가능



### 🔁 기존 방식 (argmax 기반)
| 클래스 0 (정상) | 클래스 1 (피싱) | 결과 (`argmax`)       |
| ---------- | ---------- | ------------------- |
| 0.51       | 0.49       | ✅ 정상 (0)            |
| 0.49       | 0.51       | ❌ 피싱 (1) ← 이 경우 문제!-------------------------------------------------
- 즉, 단 1% 차이만 나도 피싱이라고 판단해버렸던 것
----------------------------------------------------
### ❗ 문제점
- 확신이 낮은 상황에서도 무조건 둘 중 하나로 결정

- 예: 보이스 피싱 확률 = 0.51 → 그냥 피싱이라고 판단

- 하지만 이건 사용자 입장에서는 거짓 경고로 인식될 수 있음 (False Positive 유발)
----------------------------------------------------------
### ✅ Threshold 도입 후 (예: 0.9 기준)
| 보이스 피싱 확률 (score) | 판단 결과         |
| ----------------- | ------------- |
| 0.91              | ✅ 피싱          |
| 0.87              | ❌ 정상          |
| 0.53              | ❌ 정상 (보류/불확실- 즉, 모델이 진짜로 확신할 때만 피싱이라고 판단
-----------------------------------------------------------
## 🔍 결론
- 🔺 기존 argmax 방식: 확신도에 관계없이 높은 쪽으로 판단

- ✅ 지금 threshold 방식: 확신이 높은 경우에만 피싱 판단 → 오경고 감소) |
 |


## ✅ 목표: 기존 CSV (KorCCVi_stt_merged.csv)에 대해 threshold=0.9 적용해서 예측 결과 저장

0. 모델 구조 + 가중치 로드

In [49]:
import torch
from torch import nn
from kobert_transformers import get_kobert_model, get_tokenizer

# ✅ 토크나이저 및 BERT 모델 로딩
kobert_model = get_kobert_model()
kobert_tokenizer = get_tokenizer()

# ✅ 장치 설정
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# ✅ KoBERT 분류기 구조 정의
class KoBERTClassifier(nn.Module):
    def __init__(self, bert_model, hidden_size=768, num_classes=2, dr_rate=0.3):
        super(KoBERTClassifier, self).__init__()
        self.bert = bert_model
        self.dr_rate = dr_rate
        self.classifier = nn.Linear(hidden_size, num_classes)
        if dr_rate:
            self.dropout = nn.Dropout(p=dr_rate)
        self.softmax = nn.Softmax(dim=1)

    def forward(self, input_ids, attention_mask, token_type_ids):
        outputs = self.bert(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
        cls_output = outputs.pooler_output
        if self.dr_rate:
            cls_output = self.dropout(cls_output)
        logits = self.classifier(cls_output)
        return self.softmax(logits)

# ✅ 모델 인스턴스 생성 + 가중치 불러오기
model = KoBERTClassifier(kobert_model).to(device)
model.load_state_dict(torch.load('D:/2025_work/2025_VoicePhshing_Detection_Model/kobert_model.pt', map_location=device))
model.eval()

print("✅ KoBERT 모델 불러오기 완료")

✅ KoBERT 모델 불러오기 완료


✅ tokenizer도 함께 불러와야 함

In [43]:
kobert_tokenizer = get_tokenizer()

1. threshold=0.9 방식으로 전체 CSV 예측 수행

In [52]:
import pandas as pd

# 예측 함수 정의 (Threshold 기반)
def predict_phishing_with_threshold(text, model, tokenizer, threshold=0.9, max_len=128):
    encoded = tokenizer(
        text,
        padding='max_length',
        truncation=True,
        max_length=max_len,
        return_tensors='pt'
    )
    input_ids = encoded['input_ids'].to(device)
    attention_mask = encoded['attention_mask'].to(device)
    token_type_ids = encoded['token_type_ids'].to(device)

    with torch.no_grad():
        output = model(input_ids, attention_mask, token_type_ids)
        phishing_score = output[0][1].item()

    pred_label = 1 if phishing_score >= threshold else 0
    return pred_label, phishing_score

# 📂 병합된 CSV 로드
csv_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_merged.csv'
df = pd.read_csv(csv_path)

# 전체 예측 수행
predicted_labels = []
confidences = []

print("🔍 예측 수행 중...")

for i, text in enumerate(df['Transcript']):
    pred_label, confidence = predict_phishing_with_threshold(text, model, kobert_tokenizer, threshold=0.9)
    predicted_labels.append(pred_label)
    confidences.append(confidence)

    if (i + 1) % 100 == 0 or (i + 1) == len(df):
        print(f" - {i + 1}/{len(df)}개 처리 완료")

# 결과 저장
df['PredictedLabel'] = predicted_labels
df['Confidence'] = confidences
output_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_thresholded.csv'
df.to_csv(output_path, index=False)

print(f"\n✅ 예측 결과 저장 완료: {output_path}")

🔍 예측 수행 중...
 - 100/1418개 처리 완료
 - 200/1418개 처리 완료
 - 300/1418개 처리 완료
 - 400/1418개 처리 완료
 - 500/1418개 처리 완료
 - 600/1418개 처리 완료
 - 700/1418개 처리 완료
 - 800/1418개 처리 완료
 - 900/1418개 처리 완료
 - 1000/1418개 처리 완료
 - 1100/1418개 처리 완료
 - 1200/1418개 처리 완료
 - 1300/1418개 처리 완료
 - 1400/1418개 처리 완료
 - 1418/1418개 처리 완료

✅ 예측 결과 저장 완료: D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_thresholded.csv


In [54]:
import pandas as pd
from collections import Counter
from sklearn.metrics import classification_report

# ✅ CSV 경로 (threshold=0.9로 예측한 결과 파일)
path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_thresholded.csv'

# ✅ CSV 불러오기
df = pd.read_csv(path)

# ✅ 전체 샘플 수
total = len(df)

# ✅ 맞춘 샘플 수 계산
correct = (df['Label'] == df['PredictedLabel']).sum()
incorrect = total - correct

# ✅ 실제 라벨 / 예측 라벨 분포
true_label_dist = Counter(df['Label'])
pred_label_dist = Counter(df['PredictedLabel'])

# ✅ 정밀도, 재현율, F1-score 리포트
report = classification_report(df['Label'], df['PredictedLabel'], target_names=['정상 통화', '보이스 피싱'])

# ✅ 출력
print("📊 예측 통계 요약:")
print(f"- 총 샘플 수: {total}")
print(f"- 정확히 예측한 샘플: {correct}개 ({correct/total*100:.2f}%)")
print(f"- 잘못 예측한 샘플: {incorrect}개 ({incorrect/total*100:.2f}%)\n")

print("🔎 실제 라벨 분포:", dict(true_label_dist))
print("🔍 예측 라벨 분포:", dict(pred_label_dist))

print("\n📈 상세 성능 리포트 (정밀도, 재현율, F1-score):")
print(report)

📊 예측 통계 요약:
- 총 샘플 수: 1418
- 정확히 예측한 샘플: 1317개 (92.88%)
- 잘못 예측한 샘플: 101개 (7.12%)

🔎 실제 라벨 분포: {1: 709, 0: 709}
🔍 예측 라벨 분포: {1: 808, 0: 610}

📈 상세 성능 리포트 (정밀도, 재현율, F1-score):
              precision    recall  f1-score   support

       정상 통화       1.00      0.86      0.92       709
      보이스 피싱       0.88      1.00      0.93       709

    accuracy                           0.93      1418
   macro avg       0.94      0.93      0.93      1418
weighted avg       0.94      0.93      0.93      1418



✅ 변화 요약 (기존 vs threshold=0.9 비교)

| 항목                  | 기존 (`argmax`) | Threshold=0.9 | 변화      |
| ------------------- | ------------- | ------------- | ------- |
| **정확도 (Accuracy)**  | 92.38%        | **92.88%**    | ▲ +0.5% |
| **정상 Precision**    | 0.99          | **1.00**      | ▲ 개선됨   |
| **정상 Recall**       | 0.85          | **0.86**      | ▲ 소폭 개선 |
| **보이스피싱 Recall**    | 0.99          | **1.00**      | ▲ 개선    |
| **보이스피싱 Precision** | 0.87          | **0.88**      | ▲ 소폭 개선 |
| **오분류 수**           | 108           | **101**       | ▼ 감소    |
-------------------------------------------------------
 - 다만, 이미 모델이 꽤 잘 학습되어 있었기 때문에 threshold 조절만으로 큰 변화는 어려

1. 오분류된 정상 통화 샘플 수집 (FP 분석)

In [61]:
# 오분류된 정상 통화 (Label=0인데 PredictedLabel=1)
false_positives = df[(df['Label'] == 0) & (df['PredictedLabel'] == 1)]
false_positives.to_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/false_positives_label0_pred1.csv', index=False)
print(f"❌ 오분류된 정상 통화 샘플 수: {len(false_positives)}개 저장 완료.")

❌ 오분류된 정상 통화 샘플 수: 100개 저장 완료.


2. 이 샘플들을 포함한 보강 학습셋 구성
- Label == 0이고 PredictedLabel == 1인 정상 문장을 oversampling

- 기존 학습 데이터셋과 concat해서 보강 학습용 train.csv 생성

✅ 목표
- 기존 원본 학습 데이터셋 (KorCCVi_v1.3_fullcleansed.csv) + 오분류된 정상통화 샘플 ➕

- → 새로운 train_boosted.csv 학습 파일 생성

- 정상 라벨(0) 보강을 통해 모델이 더 신중하게 판단하도록 유도
--------------------------------------------------------------------
✅ 전제 파일 경로
- 기존 학습 데이터셋:
D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/phishing_dataset/KorCCVi_dataset/KorCCViD_v1.3_fullcleansed.csv

- 오분류된 정상통화 샘플:
D:/2025_work/2025_VoicePhshing_Detection_Model/false_positives_label0_pred1.csv

In [67]:
import pandas as pd

# ✅ 원본 학습 데이터 로드
original_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/phishing_dataset/KorCCVi_dataset/KorCCViD_v1.3_fullcleansed.csv'
df_original = pd.read_csv(original_path)[['Transcript', 'Label']]

# ✅ 오분류된 정상 통화 샘플 로드
false_positive_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/false_positives_label0_pred1.csv'
df_fp = pd.read_csv(false_positive_path)[['Transcript', 'Label']]

# ✅ 기존 데이터에 오분류 샘플 추가 (중복 제거)
df_combined = pd.concat([df_original, df_fp], ignore_index=True).drop_duplicates()

# ✅ 셔플 (무작위 섞기)
df_combined = df_combined.sample(frac=1.0, random_state=42).reset_index(drop=True)

# ✅ 저장
boosted_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/train_boosted.csv'
df_combined.to_csv(boosted_path, index=False)

print(f"✅ 보강 학습용 CSV 저장 완료: {boosted_path}")
print(f"총 샘플 수: {len(df_combined)}개 (원본 + 오분류 포함)")

✅ 보강 학습용 CSV 저장 완료: D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/train_boosted.csv
총 샘플 수: 1315개 (원본 + 오분류 포함)


3. 오분류된 정상 통화(실제로는 정상인데 피싱으로 판단했던 샘플)를 추가로 학습시켜
모델이 ‘정상 통화’에 대해 더 신중하게 판단하도록 만드는 단계
--------------------------------------------
| 항목     | 값                                  |
| ------ | ---------------------------------- |
| 데이터    | `train_boosted.csv` (1315개, 라벨 포함) |
| 모델     | KoBERT 기반 분류기 (`KoBERTClassifier`) |
| Epochs | **5\~10회** 권장 (기존보다 조금 더 반복 학습)    |
| 목적     | **오탐 줄이기 (False Positive 감소)**    |


In [73]:
# 🔧 라이브러리 임포트 (정상 작동용)
import torch
from torch import nn
from torch.utils.data import Dataset, DataLoader
from torch.optim import AdamW  # ✅ transformers 말고 torch에서 불러오기

from transformers import get_cosine_schedule_with_warmup
from kobert_transformers import get_tokenizer, get_kobert_model

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from tqdm import tqdm
from collections import Counter
from konlpy.tag import Okt

# ✅ CSV 로드
csv_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/train_boosted.csv'
df = pd.read_csv(csv_path)

# ✅ 데이터 분리
texts = df['Transcript'].tolist()
labels = df['Label'].tolist()
train_texts, val_texts, train_labels, val_labels = train_test_split(texts, labels, test_size=0.2, random_state=42)

# ✅ KoBERT 로딩
tokenizer = get_tokenizer()
bert_model = get_kobert_model()

# ✅ Dataset 클래스
class BERTDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len=128):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __getitem__(self, idx):
        encoded = self.tokenizer(
            self.texts[idx],
            padding='max_length',
            truncation=True,
            max_length=self.max_len,
            return_tensors='pt'
        )
        item = {key: val.squeeze(0) for key, val in encoded.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.texts)

# ✅ 데이터로더 생성
train_dataset = BERTDataset(train_texts, train_labels, tokenizer)
val_dataset = BERTDataset(val_texts, val_labels, tokenizer)
train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=16)

# ✅ 모델 정의
class KoBERTClassifier(nn.Module):
    def __init__(self, bert_model, hidden_size=768, num_classes=2, dr_rate=0.3):
        super(KoBERTClassifier, self).__init__()
        self.bert = bert_model
        self.dropout = nn.Dropout(p=dr_rate)
        self.classifier = nn.Linear(hidden_size, num_classes)
        self.softmax = nn.Softmax(dim=1)

    def forward(self, input_ids, attention_mask, token_type_ids):
        outputs = self.bert(input_ids=input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
        cls_output = outputs.pooler_output
        cls_output = self.dropout(cls_output)
        logits = self.classifier(cls_output)
        return self.softmax(logits)

# ✅ 모델 인스턴스 생성
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = KoBERTClassifier(bert_model).to(device)

# ✅ 옵티마이저 및 스케줄러
no_decay = ['bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
    {'params': [p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
    {'params': [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]
optimizer = AdamW(optimizer_grouped_parameters, lr=2e-5)
loss_fn = nn.CrossEntropyLoss()
EPOCHS = 5
total_steps = len(train_loader) * EPOCHS
scheduler = get_cosine_schedule_with_warmup(optimizer, num_warmup_steps=int(0.1 * total_steps), num_training_steps=total_steps)

# ✅ 학습 루프
for epoch in range(EPOCHS):
    model.train()
    total_loss, total_acc = 0, 0
    for batch in tqdm(train_loader, desc=f"Epoch {epoch+1}"):
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        token_type_ids = batch['token_type_ids'].to(device)
        labels = batch['labels'].to(device)

        preds = model(input_ids, attention_mask, token_type_ids)
        loss = loss_fn(preds, labels)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        scheduler.step()

        total_loss += loss.item()
        total_acc += (preds.argmax(1) == labels).sum().item()

    avg_loss = total_loss / len(train_loader)
    avg_acc = total_acc / len(train_dataset)
    print(f"[Epoch {epoch+1}] 평균 손실: {avg_loss:.4f} | 정확도: {avg_acc*100:.2f}%")

# ✅ 최종 모델 저장
torch.save(model.state_dict(), 'D:/2025_work/kobert_boosted.pt')
print("✅ 보강 학습된 KoBERT 모델 저장 완료")

Epoch 1: 100%|█████████████████████████████████████████████████████████████████████████| 66/66 [14:19<00:00, 13.02s/it]


[Epoch 1] 평균 손실: 0.5238 | 정확도: 83.27%


Epoch 2: 100%|█████████████████████████████████████████████████████████████████████████| 66/66 [13:53<00:00, 12.64s/it]


[Epoch 2] 평균 손실: 0.3430 | 정확도: 98.86%


Epoch 3: 100%|█████████████████████████████████████████████████████████████████████████| 66/66 [12:54<00:00, 11.73s/it]


[Epoch 3] 평균 손실: 0.3293 | 정확도: 99.05%


Epoch 4: 100%|█████████████████████████████████████████████████████████████████████████| 66/66 [13:45<00:00, 12.50s/it]


[Epoch 4] 평균 손실: 0.3251 | 정확도: 99.24%


Epoch 5: 100%|█████████████████████████████████████████████████████████████████████████| 66/66 [12:25<00:00, 11.29s/it]


[Epoch 5] 평균 손실: 0.3184 | 정확도: 99.90%
✅ 보강 학습된 KoBERT 모델 저장 완료


In [75]:
# ✅ 저장된 보강 모델 불러오기
model = KoBERTClassifier(bert_model).to(device)
model.load_state_dict(torch.load('D:/2025_work/kobert_boosted.pt', map_location=device))
model.eval()

# ✅ 예측 함수 (threshold=0.9)
def predict_phishing_threshold(text, model, tokenizer, threshold=0.9):
    encoded = tokenizer(text, padding='max_length', truncation=True, max_length=128, return_tensors='pt')
    input_ids = encoded['input_ids'].to(device)
    attention_mask = encoded['attention_mask'].to(device)
    token_type_ids = encoded['token_type_ids'].to(device)
    with torch.no_grad():
        output = model(input_ids, attention_mask, token_type_ids)
        phishing_score = output[0][1].item()
    pred_label = 1 if phishing_score >= threshold else 0
    return pred_label, phishing_score

# ✅ CSV 로드
import pandas as pd
df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_merged.csv')

# ✅ 예측 수행
predicted_labels = []
confidences = []

for i, text in enumerate(df['Transcript']):
    pred_label, confidence = predict_phishing_threshold(text, model, kobert_tokenizer, threshold=0.9)
    predicted_labels.append(pred_label)
    confidences.append(confidence)
    if (i + 1) % 100 == 0 or (i + 1) == len(df):
        print(f" - {i + 1}/{len(df)}개 처리 완료")

# ✅ 결과 저장
df['PredictedLabel'] = predicted_labels
df['Confidence'] = confidences
output_path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv'
df.to_csv(output_path, index=False)

print(f"✅ 새 모델 기반 예측 결과 저장 완료: {output_path}")


 - 100/1418개 처리 완료
 - 200/1418개 처리 완료
 - 300/1418개 처리 완료
 - 400/1418개 처리 완료
 - 500/1418개 처리 완료
 - 600/1418개 처리 완료
 - 700/1418개 처리 완료
 - 800/1418개 처리 완료
 - 900/1418개 처리 완료
 - 1000/1418개 처리 완료
 - 1100/1418개 처리 완료
 - 1200/1418개 처리 완료
 - 1300/1418개 처리 완료
 - 1400/1418개 처리 완료
 - 1418/1418개 처리 완료
✅ 새 모델 기반 예측 결과 저장 완료: D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv


In [77]:
import pandas as pd
from collections import Counter
from sklearn.metrics import classification_report

# 📥 예측 결과 CSV 로드
path = 'D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv'
df = pd.read_csv(path)

# ✅ 정확도 계산
total = len(df)
correct = (df['Label'] == df['PredictedLabel']).sum()
incorrect = total - correct

# ✅ 라벨 분포
true_label_dist = Counter(df['Label'])
pred_label_dist = Counter(df['PredictedLabel'])

# ✅ 정밀도/재현율/F1-score 계산
report = classification_report(df['Label'], df['PredictedLabel'], target_names=['정상 통화', '보이스 피싱'])

# ✅ 출력
print("📊 예측 통계 요약:")
print(f"- 총 샘플 수: {total}")
print(f"- 정확히 예측한 샘플: {correct}개 ({correct/total*100:.2f}%)")
print(f"- 잘못 예측한 샘플: {incorrect}개 ({incorrect/total*100:.2f}%)\n")

print("🔎 실제 라벨 분포:", dict(true_label_dist))
print("🔍 예측 라벨 분포:", dict(pred_label_dist))

print("\n📈 상세 성능 리포트 (정밀도, 재현율, F1-score):")
print(report)

📊 예측 통계 요약:
- 총 샘플 수: 1418
- 정확히 예측한 샘플: 1318개 (92.95%)
- 잘못 예측한 샘플: 100개 (7.05%)

🔎 실제 라벨 분포: {1: 709, 0: 709}
🔍 예측 라벨 분포: {1: 609, 0: 809}

📈 상세 성능 리포트 (정밀도, 재현율, F1-score):
              precision    recall  f1-score   support

       정상 통화       0.88      1.00      0.93       709
      보이스 피싱       1.00      0.86      0.92       709

    accuracy                           0.93      1418
   macro avg       0.94      0.93      0.93      1418
weighted avg       0.94      0.93      0.93      1418



In [81]:
import pandas as pd

# 📥 예측 결과 CSV 로드
df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv')

# ✅ 컬럼 목록 확인
print("📌 CSV 컬럼 목록:")
print(df.columns.tolist())

📌 CSV 컬럼 목록:
['Label', 'Transcript', 'Filename', 'PredictedLabel', 'Confidence']


In [83]:
import pandas as pd

# 📥 예측 결과 CSV 로드 (보강 학습 모델 기준)
df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv')

# ❌ 오분류된 샘플만 추출
misclassified = df[df['Label'] != df['PredictedLabel']]

# 🎯 FP: 정상인데 피싱으로 예측
false_positives = misclassified[(misclassified['Label'] == 0) & (misclassified['PredictedLabel'] == 1)]

# 🎯 FN: 피싱인데 정상으로 예측
false_negatives = misclassified[(misclassified['Label'] == 1) & (misclassified['PredictedLabel'] == 0)]

# ✅ 개수 출력
print(f"❌ 오분류 샘플 총합: {len(misclassified)}")
print(f"🔺 False Positives (정상 → 피싱): {len(false_positives)}")
print(f"🔻 False Negatives (피싱 → 정상): {len(false_negatives)}\n")

# ✅ 예시 출력 (앞부분만)
print("📋 FP 예시:")
print(false_positives[['Filename','Transcript', 'Confidence']].head(5))

print("\n📋 FN 예시:")
print(false_negatives[['Filename','Transcript', 'Confidence']].head(5))


❌ 오분류 샘플 총합: 100
🔺 False Positives (정상 → 피싱): 0
🔻 False Negatives (피싱 → 정상): 100

📋 FP 예시:
Empty DataFrame
Columns: [Filename, Transcript, Confidence]
Index: []

📋 FN 예시:
                       Filename  \
1055                   KorCCViD   
1318   phishing_수사기관_사칭형_87.wav   
1319   phishing_수사기관_사칭형_90.wav   
1320  phishing_수사기관_사칭형_130.wav   
1321  phishing_수사기관_사칭형_173.wav   

                                             Transcript  Confidence  
1055                                그리고 드리 더라도 예뻐 아니 걸로    0.015059  
1318  예. 남순씨가 어떤 사람인가요? 그 중에 선생님 명의로 대신 SC 스탠라드 은혜 통...    0.020113  
1319  네. 아니요. 그러시라면서 혹시 저 ***린 셰이*** 남성분인데요 ***린 사람들...    0.023652  
1320  네. 지갑이나 신분증, 여권 등을 분실하신 적은 있으십니까? 없어요. 없는 것 같은...    0.017453  
1321  네? 여보세요? 네? 다름이 아니라 일로 된 명의돈의 사건 하나가 접수가 돼서 여기...    0.015759  


In [85]:
import pandas as pd

# 예측 결과 불러오기
df = pd.read_csv('D:/2025_work/2025_VoicePhshing_Detection_Model/dataset/KorCCVi_stt_predicted_boosted.csv')

# 오분류 샘플 중 FN (보이스 피싱인데 정상 통화로 판단)
false_negatives = df[(df['Label'] == 1) & (df['PredictedLabel'] == 0)]

# Filename에 "KorCCVi"가 **포함되지 않은** FN 샘플 수
non_korccvid_fn = false_negatives[~false_negatives['Filename'].str.contains("KorCCVi")]

print(f"❌ 전체 False Negative 수: {len(false_negatives)}")
print(f"📂 KorCCVi 외 출처 FN 수: {len(non_korccvid_fn)}")

# 예시 출력
print("\n📋 예시 파일명:")
print(non_korccvid_fn['Filename'].head(5).to_list())


❌ 전체 False Negative 수: 100
📂 KorCCVi 외 출처 FN 수: 99

📋 예시 파일명:
['phishing_수사기관_사칭형_87.wav', 'phishing_수사기관_사칭형_90.wav', 'phishing_수사기관_사칭형_130.wav', 'phishing_수사기관_사칭형_173.wav', 'phishing_대출사기형_105.wav']


### 🧠 현재 문제 요약
| 항목              | 설명                                  |
| --------------- | ----------------------------------- |
| 🔎 현재 모델 학습 데이터 | **KorCCVi 데이터셋**만 사용 (정제된 고품질 텍스트)  |
| 🧪 테스트 대상       | KorCCVi + **직접 STT한 실환경 음성 텍스트** 혼합 |
| ❌ 오분류 FN 분석 결과  | **거의 전부 STT 데이터**에서 발생 (99/100개)    |
